### PR TITLE
feat(account deletion): Enqueue saved items for deletion

### PR DESCRIPTION
**Pull request by kkyeboah**
## Goal
Add an endpoint to enqueue `itemIds` to be deleted for a given user

## I'd love feedback/perspectives on:
- Implementation

## Implementation Decisions
- set the query limit to `5000` to reduce requests to the database to a given user
- chunk itemIds into batches of `1000`
- set the SQS message batch size to `10` - send messages in batches to reduce network calls to SQS

## References

Jira ticket: INFRA-570

### CODE REVIEW

File: src/dataService/savedItemsService.ts [hunk 1]
```diff
@@ -393,6 +393,20 @@ export class SavedItemDataService {
     return this.listItemUpdateBuilder().where('item_id', itemId);
   }
 
+  /**
+   * Get saved item IDs for a given user.
+   * @param offset
+   * @param limit
+   */
+  public getSavedItemIds(offset: number, limit: number) {
+    return this.db('list')
```
Comment (kschelonka) [inline src/dataService/savedItemsService.ts:402]: related/unrelated -- I like the idea of having a factory pattern to make all our queries safer, which always includes `andWhere('user_id', this.userId)`

Comment (kschelonka) [reply src/dataService/savedItemsService.ts:402]: don't change anything rn, just wanted to write down a thought lol

File: src/server/routes/queueDelete.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+ */
+function convertToSqsEntry(message: SqsMessage): SendMessageBatchRequestEntry {
+  return {
+    Id: nanoid(),
+    MessageBody: JSON.stringify(message),
+  };
+}
+
+/**
+ * Generate and return a list of a given size
```
Comment (kschelonka) [inline src/server/routes/queueDelete.ts:164]: I think this docstring is wrong -- more like you're yielding chunks of a list vs. generating a list of specific size

File: src/server/routes/queueDelete.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+/**
+ * Send messages in a batch to SQS
+ * @param entries
+ * @param traceId
+ */
+async function sqsSendBatch(
+  entries: SendMessageBatchRequestEntry[],
+  traceId: string
```
Comment (kschelonka) [inline src/server/routes/queueDelete.ts:142]: Unused?

File: src/server/routes/queueDelete.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+      nextChunk = chunkedIds.next();
+    }
+
+    offset += offset + limit;
+  }
+
+  // If there's any remaining, send to SQS
+  if (sqsEntries.length) {
+    sqsSends.push(sqsSendBatch(sqsEntries, traceId));
```
Comment (kschelonka) [inline src/server/routes/queueDelete.ts:122]: the trace id should probably be unique per message, not batch

File: src/server/routes/queueDelete.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    const chunkedIds = chunk(ids, config.queueDelete.itemIdChunkSize);
+
+    let nextChunk = chunkedIds.next();
+    while (!nextChunk.done) {
+      sqsEntries.push(
+        convertToSqsEntry({
+          userId,
+          email,
+          status,
+          itemIds: nextChunk.value,
```
Comment (kschelonka) [inline src/server/routes/queueDelete.ts:105]: I'd add the traceId here

File: src/server/routes/queueDelete.ts [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  savedItemService: SavedItemDataService,
+  traceId: string
+): Promise<void> {
+  const { userId, email, status } = data;
+  const limit = config.queueDelete.queryLimit;
+  let offset = 0;
+  const sqsSends: Promise<SendMessageBatchCommandOutput>[] = [];
+  let sqsEntries: SendMessageBatchRequestEntry[] = [];
+
+  const loopCondition = true;
```
Comment (kschelonka) [inline src/server/routes/queueDelete.ts:91]: this never gets updated

File: src/test/integration/savedItemsService.integration.ts [hunk 7]
```diff
@@ -79,4 +79,36 @@ describe('SavedItemsService', () => {
     expect(savedItems[0].url).to.equal('https://abc');
     expect(savedItems[1].url).to.equal('https://def');
   });
+
```
Comment (kschelonka) [inline src/test/integration/savedItemsService.integration.ts:82]: Just in case, would be nice to have the edge case where your offset is past the end of the list, since your while condition depends on that

File: src/server/routes/queueDelete.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+export type SqsMessage = {
+  userId: number;
+  email: string;
+  status: 'FREE' | 'PREMIUM';
+  itemIds: number[];
+};
+
+const router = Router();
+
+const batchDeleteSchema: Schema = {
```
Comment (kschelonka) [inline src/server/routes/queueDelete.ts:26]: copy-paste error: should probably be queueDeleteSchema?

File: src/server/routes/queueDelete.ts (hunk 6 above)
Comment (kkyeboah) [reply src/server/routes/queueDelete.ts:91]: ESlint yells at me when I just put `true` in the while loop

**Review by kschelonka — APPROVED** [review]

**Review by kschelonka — APPROVED** [review]
